File: BinarySearch.js
/*
  Binary Search Assumes PreSorted array
  Doesn't do type validation
  Non-Recursive Version
*/
function binarySearch(val, array) {
  var mid = null;
  var floor = 0;
      ceil = array.length;
  var midPoint = Math.floor((ceil+floor)/2);

  while (mid !== val){
    var midPoint = Math.floor((ceil+floor)/2);
    
    mid = array[midPoint];
    
    //Break loop when you've run out of space to check
    if(ceil === floor || (ceil - floor) === 1 ){ break; }
    
    if (val > mid){
      floor = midPoint;
    } else if(val < mid){
      ceil = midPoint;
    }
  }
  
  //Checked value other wise return null.
  return (mid === val)? mid : null;
};

/*
  Binary Search Assumes PreSorted array
  Doesn't do type validation
  Recursive Version
*/
function recursiveBinarySearch(val, array) {
  var floor = 0;
      ceil = array? array.length : 0;

  var midPoint = Math.floor((ceil+floor)/2);
  mid = array[midPoint];

  if(ceil === floor || (ceil - floor) === 1 ){
    return (mid === val)? mid : null;
  }

  if(mid === val){
    return mid;  
  } else {
    if (val > mid) {
      floor = midPoint;
    } else if(val < mid) {
      ceil = midPoint;
    }
    return recursiveBinarySearch(val, array.slice(floor , ceil));
  }
};

File: BinarySearch.swift
func binarySearch(val:Int , array:[Int]) -> Int? {
    var floor:Int = 0
    var ceil:Int = array.count
    
    var midPoint:Int? = nil //Get the Floor of this value
    var mid:Int? = nil
    
    while (mid != val) {
        midPoint = (ceil + floor)/2
        mid = array[midPoint!]

        if(ceil == floor || (ceil-floor) == 1){
           break
        }
        
        
        if (val > mid) {
            floor = midPoint!;
        } else if(val < mid) {
            ceil = midPoint!;
        }
    }
    
    if (mid == val) {
        return mid
    }
    
    return nil
}

func recursiveBinarySearch(val:Int , array:[Int]) -> Int? {
    var floor:Int = 0
    var ceil:Int = array.count
    
    let midPoint = (ceil-floor)/2
    let mid = array[midPoint];
    
    
    if(mid == val){
        return mid
    } else if(ceil == floor || (ceil - floor) == 1 ){
        return nil
    } else {
        if (val > mid) {
            floor = midPoint
        } else if(val < mid) {
            ceil = midPoint
        }
    }
    
    let newArray:[Int] = Array(array[floor..<ceil])
    return recursiveBinarySearch(val, array: newArray)
}
